src/passes/function_passes/CheckPointPass.cpp

```diff
@@ -0,0 +1,200 @@
+ * ignored.
+ */
+struct CheckPointPass: public FunctionPass {
+  static char id;
+  /*
+   * The unique identifier of the next stackmap call.
+   */
+  static uint64_t sm_id;
+
+  /*
+   * Map the function name to the ID of the first stackmap/patchpoint call
+   * in the function. This is used to ensure the runtime can associate the
+   * stackmap calls inside each function with stackmap calls from the
+   * unoptimized version of the function
+   */
+  static std::map<StringRef, uint64_t> start_sm_id;
+
+  CheckPointPass() : FunctionPass(id) {}
+
+  virtual bool runOnFunction(Function &fun) {
+    StringRef fun_name = fun.getName();
+    // This pass currently only operates on two functions: `trace` and
+    // `__unopt_trace`. This is because it is supposed to demonstrate how
+    // the runtime can resume the execution of `trace` in `__unopt_trace`.
+    if (!fun_name.endswith(TRACE_FUN_NAME)) {
+      return false;
+    }
+    outs() << "Running CheckPointPass on function: " << fun.getName() << '\n';
+
+    // Each function has a corresponding unoptimized version, which starts
+    // with the `__unopt_` prefix (the 'twin').
+    StringRef twin_fun = getTwinName(fun_name);
+
+    // The next stackmap ID to assign
+    uint64_t curr_sm_id = sm_id;
+    if (start_sm_id.find(twin_fun) != start_sm_id.end()) {
+      // If the 'twin' of this function has already been processed, index
+      // stackmap calls starting from ~ID, where ID is the identifier of the
+      // first stackmap call in the 'twin'
+      curr_sm_id = ~start_sm_id[twin_fun];
+    }
+
+    Module *mod = fun.getParent();
+    DataLayout data_layout(mod);
+    // sm_id is the ID of the first stackmap call in this function
+    start_sm_id[fun_name] = sm_id;
+    uint64_t call_inst_count = 0;
+
+    Type *i8ptr_t = PointerType::getUnqual(
+        IntegerType::getInt8Ty(mod->getContext()));
+    // The callback to call when a guard fails
+    Constant* gf_handler_ptr = ConstantExpr::getBitCast(
+        mod->getFunction(GUARD_FUN_NAME), i8ptr_t);
+    for (auto &bb : fun) {
+      for (BasicBlock::iterator it = bb.begin(); it != bb.end(); ++it) {
+        if (isa<CallInst>(it) &&
+            cast<CallInst>(*it).getCalledFunction()->getName() == "putchar") {
+          IRBuilder<> builder(&bb, it);
+          uint64_t patchpoint_id;
+          if (start_sm_id.find(twin_fun) == start_sm_id.end()) {
+            // Each stackmap ID is calculated by adding the offset
+            // (call_inst_count) to the 'base' ID (curr_sm_id)
+            patchpoint_id = curr_sm_id + call_inst_count;
+          } else {
+            // If the 'twin' of this function has already been processed,
+            // make sure the ID of each stackmap call in this function can be
+            // obtained from an ID of a stackmap call in the 'twin', by
+            // negating the id.
+            patchpoint_id = ~(~curr_sm_id + call_inst_count);
```

Comment:
I'm not following this negation logic.

As I understand, you are defining a mapping between the ids of the optimised and unoptimised functions.

How about comment with an example? Tell us what `~` does on a patch point.

Reply:
The part which calculates the next patch point ID is pretty badly written, but I have no idea how to make it clearer. It is needed, because when a guard fails in `trace` (the `patchpoint` intrinsic calls `__guard_failure`), I need to work out which stack map in `__unopt_trace` corresponds to the `patchpoint` call in `trace`. 

In `__guard_failure`, I have access to the ID of the `patchpoint` call. I have assigned patch point IDs in such a way that for any ID `sm_id`, I can obtain the ID of the corresponding `stackmap` call in the unoptimized version of the function by calculating `~sm_id`.